### ISSUE TITLE
Enable dynamic index argument for slice insert and slice remove

### ISSUE THREAD

**Issue by vezenovm** [noir-lang/noir#2462]
### Problem

Support for dynamic indices is included in this PR (https://github.com/noir-lang/noir/pull/2446). However, it still has a missing feature with handling dynamic indices for the SliceInsert and SliceRemove intrinsics. Implementing SliceInsert and SliceRemove with non-const inputs is not as straightforward as the other slice intrinsics in ACIR gen. 

A more involved codegen will be required. In both SSA gen and ACIR gen, slice intrinsicssimplifications could be performed in most cases with  the `im::Vector` builtins. Without a constant index, the `im::Vector` builtin cannot be used, and an equivalent codegen of the `insert` and `remove` methods will most likely be required.

### Happy Case

We can use insert and remove on slices with witness values.

For example, this should be enabled:
```
fn main(x: Field) {
    let mut slice = [];
    for i in 0..5 {
        slice = slice.push_back(i);
    }
    
    slice = slice.insert(x - 2, 20);
    assert( ... );
}

```

### Alternatives Considered

If this is not implemented the feature will just be missing. 

### Would you like to submit a PR for this Issue?

No

**Comment by jfecher** [comment]
I think an approach I'd like for this bug and one direction I strongly want to move towards in the future is to have a compiler error for unimplemented features like this so that users do not randomly encounter these crashes.

For example, for here I think we can add a `std::assert_constant(index)` within our slice insert and remove methods in the standard library. This should give users a nicer error while we're trying to implement it in the meantime.

**Comment by vezenovm** [comment]
This was closed as per #3624
